scripts/pylib/twister/twisterlib.py

```diff
@@ -3346,127 +3074,398 @@ def add_configurations(self):
+                        ts.depends_on = ts_dict["depends_on"]
+                        ts.min_flash = ts_dict["min_flash"]
+                        ts.extra_sections = ts_dict["extra_sections"]
+                        ts.integration_platforms = ts_dict["integration_platforms"]
+                        ts.seed = ts_dict["seed"]
+
+                        testcases = ts_dict.get("testcases", [])
+                        if testcases:
+                            for tc in testcases:
+                                ts.add_testcase(name=f"{name}.{tc}")
+                        else:
+                            # only add each testcase once
+                            for sub in set(subcases):
+                                name = "{}.{}".format(ts.id, sub)
+                                ts.add_testcase(name)
+
+                            if not subcases:
+                                ts.add_testcase(ts.id, freeform=True)
+
+                        ts.ztest_suite_names = ztest_suite_names
+
+                        if testsuite_filter:
+                            if ts.name and ts.name in testsuite_filter:
+                                self.testsuites[ts.name] = ts
+                        else:
+                            self.testsuites[ts.name] = ts
+
+                except Exception as e:
+                    logger.error("%s: can't load (skipping): %s" % (ts_path, e))
+                    self.load_errors += 1
+        return len(self.testsuites)
+
+
+
+    def scan_file(self, inf_name):
+        regular_suite_regex = re.compile(
+            # do not match until end-of-line, otherwise we won't allow
+            # stc_regex below to catch the ones that are declared in the same
+            # line--as we only search starting the end of this match
+            br"^\s*ztest_test_suite\(\s*(?P<suite_name>[a-zA-Z0-9_]+)\s*,",
+            re.MULTILINE)
+        registered_suite_regex = re.compile(
+            br"^\s*ztest_register_test_suite"
+            br"\(\s*(?P<suite_name>[a-zA-Z0-9_]+)\s*,",
+            re.MULTILINE)
+        new_suite_regex = re.compile(
+            br"^\s*ZTEST_SUITE\(\s*(?P<suite_name>[a-zA-Z0-9_]+)\s*,",
+            re.MULTILINE)
+        # Checks if the file contains a definition of "void test_main(void)"
+        # Since ztest provides a plain test_main implementation it is OK to:
+        # 1. register test suites and not call the run function iff the test
+        #    doesn't have a custom test_main.
+        # 2. register test suites and a custom test_main definition iff the test
+        #    also calls ztest_run_registered_test_suites.
+        test_main_regex = re.compile(
+            br"^\s*void\s+test_main\(void\)",
+            re.MULTILINE)
+        registered_suite_run_regex = re.compile(
+            br"^\s*ztest_run_registered_test_suites\("
+            br"(\*+|&)?(?P<state_identifier>[a-zA-Z0-9_]+)\)",
+            re.MULTILINE)
```

Comment:
@nashif I am lost now. 
In #46052 you wrote
> This PR should work on generating the test plan and not as a post step to verify the test plan and set status. We could make this work on-demand by adding an option to the yaml file to generate list of test cases only on select test suites while we continue to parse things as before on everything else.

My understanding was that we will stay with mainly relying on parsing c files with twister to get a testplan fast by default, and use refine method e.g. based on .elf parsing if the "dumb" parser fails. 
The problem with parsing .elf is that you need this .elf, hence a successful build for that, but we want to know what will be tested (having test plan) before we start testing (building).

Reply:
thanks for clarification. I have to admit I still don't understand, since for me the other way round seems more logical. Or maybe when we talk about parser we think about two different things. When I am writing parser, by default I mean a parser in twister, which is prone to final linker results pitfalls. IMO, since this is fast and doesn't require builds to finish, it can be used as the default option for getting test cases in the test plan. However, for special cases, where the tests in the binary doesn't correspond to the test plan, we can tell twister to align the plan based on the content of the binary (or some other method, e.g. explicitly adding testcases to yaml). Such cases could be exposed in advance thanks to #46052 and #46278. 
But anyway, I think we have the same goal in mind, i.e. reliable test plan with test cases. Maybe during a testing wg call I'll be able to better understand how we can get there.

Reply:
I would really love to understand this better. I'm not sure I'm seeing a benefit of having the test plan before the binary is built. Aaron will be attending the next testing working group meeting and I've asked to make sure that our testing team is aligned on this and we understand fully what the goal/benefit is.

Reply:
@yperess I am on the fence with this one. This was originally brought up here https://github.com/zephyrproject-rtos/zephyr/issues/28864 long time ago, but as we are all aware, it was and still  in most cases a testsuite configuration/setup problem rather than an issue with twister.

The only issue I see with this is when you have tests that are not building for whatever reason, in this case we will not be able to extract testcases. 

I do prefer where all testsuites/testcases are extracted in a consistent and complete way rather than depend on parsing like we do now. So if we figure out how to deal with such cases, I think we should be good.